---
output: 
  html_document: 
    toc: yes
    toc_depth: 4
    toc_float: yes
--- 


# sc10x analysis                  
            
         
##### load dependancies                
```{r message=FALSE, warning=FALSE, include=FALSE}
source("I:/Shared_win/projects/RNA_normal/analysis.10x.r")
```
       
        
Baf53cre Ahr-cko vs ctl,           
CR infection,         
Intestinal RORgt+ Immune Cells         
            
         
loading 72,000 for all                     
cellranger calling 34,620 cells, mean reads 11,163            
       
       
t0 initial             
        
       
## load 10x data      

```{r}
filt.10x <- Read10X(data.dir = "I:/Shared_win/projects/20230927_10x_SZJ/output_demo/filtered_feature_bc_matrix/")
```

```{r}
GEX <- filt.10x$`Gene Expression`
FB <- filt.10x$`Antibody Capture`
```


### check datasets   
```{r message=FALSE, warning=FALSE}
dim(GEX)
GEX[1:6,1:6]
```


```{r}
dim(FB)
FB[,1:6]
```

```{r}
rowSums(FB)
```

```{r}
rowSums(FB>0)
```


## FB     
      
```{r}
# build seurat object
FB.seur <- CreateSeuratObject(counts = FB,project = "ILC3_FB")

FB.seur
```

```{r}
rownames(FB.seur)
```


### normalization     
    
perform Seurat 'Demultiplexing with hashtag oligos(HTOs)     
ref https://satijalab.org/seurat/articles/hashing_vignette.html       
    
```{r}
# normalize
FB.seur <- NormalizeData(FB.seur)
#FB.seur <- FindVariableFeatures(FB.seur, selection.method = "mean.var.plot")
FB.seur <- ScaleData(FB.seur, features = VariableFeatures(FB.seur))

# independent assay for HTO data  
FB.seur[['HTO']] <- CreateAssayObject(counts = FB)
FB.seur <- NormalizeData(FB.seur, assay = 'HTO', normalization.method = 'CLR')
```


### check demux         

first define FB colors based on conditions                      

```{r fig.height=6, fig.width=8.1}
#scales::show_col(ggsci::pal_igv("default")(49))
```


```{r fig.height=5, fig.width=5}
#scales::show_col(ggsci::pal_d3("category20b")(20))
#scales::show_col(ggsci::pal_d3("category20c")(20))
```


```{r}
#
color.FB <- ggsci::pal_d3("category20c")(20)[c(2,7,12,17,
                                               1,6,11,16
                                               )]


level.FB <- c(rownames(FB.seur))

color.FBraw <-  c("#FF6C91","lightgrey",color.FB)
```

```{r}
scales::show_col(color.FBraw, ncol = 4)
scales::show_col(color.FB, ncol = 4)
```



```{r fig.height=6.5, fig.width=12, message=FALSE, warning=FALSE}
par(mfrow=c(2,4))

#plot(sort(rowSums(t(FB))),
#     xlab="reordered cells", 
#     ylab="UMI counts", log="xy",
#     main="sum")


plot(sort(t(FB)[,1]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[1])
plot(sort(t(FB)[,2]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[2])
plot(sort(t(FB)[,3]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[3])
plot(sort(t(FB)[,4]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[4])
plot(sort(t(FB)[,5]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[5])
plot(sort(t(FB)[,6]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[6])
plot(sort(t(FB)[,7]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[7])
plot(sort(t(FB)[,8]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[8])

```


#### range 'q'         

##### q0.50 ~ 0.99       

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
FB.test <- FB.seur
#cut.list <- list()

table.demux <- data.frame(sample=c("quantile","Doublet",rownames(FB.test),"Negative"))

for(i in 50:99){
  FB.test <- HTODemux(FB.test, assay = "HTO", positive.quantile = i/100)
  table.demux <- cbind(table.demux,
                       c(i/100,table(FB.test$hash.ID)[c("Doublet",rownames(FB.test),"Negative")]))
}
rownames(table.demux) <- table.demux$sample
table.demux <- table.demux[,2:ncol(table.demux)]
table.demux <- data.frame(t(table.demux))
rownames(table.demux) <- NULL
#table.demux
```

```{r fig.width=13.5, fig.height=10}
#color.FB <- scales::hue_pal()(10)
par(mfrow=c(3,4))


plot(table.demux$Doublet, type="p", col="#FF6C91", pch=16, ylab="Doublet")
plot(table.demux$Negative, type="p", col="lightgrey", pch=16, ylab="Negative")
plot(rowSums(table.demux[,grep("CKO|CTL",colnames(table.demux))]), type="p", col="#306000", pch=16, ylab="Singlet")

plot(table.demux$quantile, pch=16, ylab="mod-quantile")
#plot.new()

plot(table.demux[,2+1], type="p", col=color.FB[1], pch=16, ylab=colnames(table.demux)[2+1])
plot(table.demux[,2+2], type="p", col=color.FB[2], pch=16, ylab=colnames(table.demux)[2+2])
plot(table.demux[,2+3], type="p", col=color.FB[3], pch=16, ylab=colnames(table.demux)[2+3])
plot(table.demux[,2+4], type="p", col=color.FB[4], pch=16, ylab=colnames(table.demux)[2+4])

plot(table.demux[,2+5], type="p", col=color.FB[5], pch=16, ylab=colnames(table.demux)[2+5])
plot(table.demux[,2+6], type="p", col=color.FB[6], pch=16, ylab=colnames(table.demux)[2+6])
plot(table.demux[,2+7], type="p", col=color.FB[7], pch=16, ylab=colnames(table.demux)[2+7])
plot(table.demux[,2+8], type="p", col=color.FB[8], pch=16, ylab=colnames(table.demux)[2+8])

```


##### q0.9900 ~ 0.9999               

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
FB.test <- FB.seur
#cut.list <- list()

table.demux.s <- data.frame(sample=c("quantile","Doublet",rownames(FB.test),"Negative"))

for(i in 9900:9999){
  FB.test <- HTODemux(FB.test, assay = "HTO", positive.quantile = i/10000)
  table.demux.s <- cbind(table.demux.s,
                       c(i/10000,table(FB.test$hash.ID)[c("Doublet",rownames(FB.test),"Negative")]))
}
rownames(table.demux.s) <- table.demux.s$sample
table.demux.s <- table.demux.s[,2:ncol(table.demux.s)]
table.demux.s <- data.frame(t(table.demux.s))
rownames(table.demux.s) <- NULL
#table.demux.s
```


```{r fig.width=13.5, fig.height=10}
#color.FB <- scales::hue_pal()(10)
par(mfrow=c(3,4))


plot(table.demux.s$Doublet, type="p", col="#FF6C91", pch=16, ylab="Doublet")
plot(table.demux.s$Negative, type="p", col="lightgrey", pch=16, ylab="Negative")
plot(rowSums(table.demux.s[,grep("CKO|CTL",colnames(table.demux.s))]), type="p", col="#306000", pch=16, ylab="Singlet")

plot(table.demux.s$quantile, pch=16, ylab="mod-quantile")
#plot.new()

plot(table.demux.s[,2+1], type="p", col=color.FB[1], pch=16, ylab=colnames(table.demux.s)[2+1])
plot(table.demux.s[,2+2], type="p", col=color.FB[2], pch=16, ylab=colnames(table.demux.s)[2+2])
plot(table.demux.s[,2+3], type="p", col=color.FB[3], pch=16, ylab=colnames(table.demux.s)[2+3])
plot(table.demux.s[,2+4], type="p", col=color.FB[4], pch=16, ylab=colnames(table.demux.s)[2+4])

plot(table.demux.s[,2+5], type="p", col=color.FB[5], pch=16, ylab=colnames(table.demux.s)[2+5])
plot(table.demux.s[,2+6], type="p", col=color.FB[6], pch=16, ylab=colnames(table.demux.s)[2+6])
plot(table.demux.s[,2+7], type="p", col=color.FB[7], pch=16, ylab=colnames(table.demux.s)[2+7])
plot(table.demux.s[,2+8], type="p", col=color.FB[8], pch=16, ylab=colnames(table.demux.s)[2+8])

```

### demultiplexing        

    
```{r}
# Demultiplex cells based on HTO enrichment
FB.seur <- HTODemux(FB.seur, assay = 'HTO', positive.quantile = 0.99)

table(FB.seur$HTO_classification.global)
table(FB.seur$hash.ID)[c("Doublet","Negative",rownames(FB))]
```


```{r}
# Demultiplex cells based on HTO enrichment
FB.seur <- HTODemux(FB.seur, assay = 'HTO', positive.quantile = 0.992)

table(FB.seur$HTO_classification.global)
table(FB.seur$hash.ID)[c("Doublet","Negative",rownames(FB))]
```


```{r}
# Demultiplex cells based on HTO enrichment
FB.seur <- HTODemux(FB.seur, assay = 'HTO', positive.quantile = 0.994)

table(FB.seur$HTO_classification.global)
table(FB.seur$hash.ID)[c("Doublet","Negative",rownames(FB))]
```


```{r}
# Demultiplex cells based on HTO enrichment
FB.seur <- HTODemux(FB.seur, assay = 'HTO', positive.quantile = 0.996)

table(FB.seur$HTO_classification.global)
table(FB.seur$hash.ID)[c("Doublet","Negative",rownames(FB))]
```


```{r}
# Demultiplex cells based on HTO enrichment
FB.seur <- HTODemux(FB.seur, assay = 'HTO', positive.quantile = 0.998)

table(FB.seur$HTO_classification.global)
table(FB.seur$hash.ID)[c("Doublet","Negative",rownames(FB))]
```



```{r}
## tags q0.99
#cutoff.FB <- c(85,237,46,121,
#               23,160,86,92)
## custom
# ref
# CKO.1  q0.996  102
# CKO.2  q0.99  237
# CKO.3  q0.998  55
# CKO.4  q0.99  121
# CTL.1  q0.994  25
# CTL.2  q0.99  160
# CTL.3  q0.996  102
# CTL.4  q0.996  112
#
cutoff.FB <- c(102,237,55,121,
               25,160,102,112)

names(cutoff.FB) <- rownames(FB.seur)
cutoff.FB
```

```{r paged.print=FALSE}
data.frame(cutoff.FB=cutoff.FB)
```

```{r fig.height=6.5, fig.width=12, message=FALSE, warning=FALSE}
par(mfrow=c(2,4))

#plot(sort(rowSums(t(FB))),
#     xlab="reordered cells", 
#     ylab="UMI counts", log="xy",
#     main="sum")

plot(sort(t(FB)[,1]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[1])
abline(h=cutoff.FB[1],col = color.FB[1])
plot(sort(t(FB)[,2]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[2])
abline(h=cutoff.FB[2],col = color.FB[2])
plot(sort(t(FB)[,3]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[3])
abline(h=cutoff.FB[3],col = color.FB[3])
plot(sort(t(FB)[,4]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[4])
abline(h=cutoff.FB[4],col = color.FB[4])
plot(sort(t(FB)[,5]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[5])
abline(h=cutoff.FB[5],col = color.FB[5])

plot(sort(t(FB)[,6]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[6])
abline(h=cutoff.FB[6],col = color.FB[6])
plot(sort(t(FB)[,7]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[7])
abline(h=cutoff.FB[7],col = color.FB[7])
plot(sort(t(FB)[,8]),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[8])
abline(h=cutoff.FB[8],col = color.FB[8])

```


```{r fig.height=6.5, fig.width=12, message=FALSE, warning=FALSE}
par(mfrow=c(2,4))

#plot(sort(rowSums(t(FB))),
#     xlab="reordered cells", 
#     ylab="UMI counts", log="xy",
#     main="sum")

plot(sort(t(FB)[,1],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[1])
abline(h=cutoff.FB[1],col = color.FB[1])
plot(sort(t(FB)[,2],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[2])
abline(h=cutoff.FB[2],col = color.FB[2])
plot(sort(t(FB)[,3],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[3])
abline(h=cutoff.FB[3],col = color.FB[3])
plot(sort(t(FB)[,4],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[4])
abline(h=cutoff.FB[4],col = color.FB[4])
plot(sort(t(FB)[,5],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[5])
abline(h=cutoff.FB[5],col = color.FB[5])

plot(sort(t(FB)[,6],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[6])
abline(h=cutoff.FB[6],col = color.FB[6])
plot(sort(t(FB)[,7],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[7])
abline(h=cutoff.FB[7],col = color.FB[7])
plot(sort(t(FB)[,8],decreasing = T),
     xlab="reordered cells", 
     ylab="UMI counts", log="xy",
     main=rownames(FB)[8])
abline(h=cutoff.FB[8],col = color.FB[8])

```


```{r}
## custom cutoff run this
custom.Demux <- function(FBobj,FBcutoff){
  # define decision matrix
  dd <- FBobj@assays[['HTO']]@counts
  dd <- apply(dd, 2, function(x){
    x > FBcutoff
  })
  x1 <- colSums(dd)
  x1[x1>1] <- "Doublet"
  x1[x1==0] <- "Negative"
  x1[x1==1] <- "Singlet"

  x2 <- x1
  
  bc.slt  <- names(x1)[x1=="Singlet"]
  
  for(bc in bc.slt){
    x2[bc] <- rownames(dd)[dd[,bc]]
  }
  
  FBdf <- data.frame(HTO_classification.global=factor(x1,
                                                      levels = c("Doublet","Negative","Singlet")),
                     hash.ID=factor(x2,
                                    levels = c("Doublet","Negative",rownames(dd))))
  return(FBdf)
}

df.FB <- custom.Demux(FB.seur,cutoff.FB)
```


```{r paged.print=FALSE}
## custom cutoff run this
dim(df.FB)
df.FB[1:10,]
```

```{r}
## custom cutoff run this
table(df.FB)
```

```{r paged.print=FALSE}
## custom cutoff run this
FB.seur@meta.data[,c("HTO_classification.global","hash.ID")] <- df.FB
FB.seur@meta.data[1:4,]
```

```{r}
## default q0.99 run this
#FB.seur$HTO_classification.global <- factor(as.character(FB.seur$HTO_classification.global),
#                                            levels = c("Doublet","Negative","Singlet"))
#FB.seur$hash.ID <- factor(as.character(FB.seur$hash.ID),
#                          levels = c("Doublet","Negative",rownames(FB.seur)))
```


```{r singlelt, fig.height=4, fig.width=4}
sl_stat <- table(FB.seur$HTO_classification.global)
barplot(sl_stat,ylim = c(0,28500),col = c("#FF6C91","lightgrey","#7CAE00"),main = "Feature Barcode statistics")
text(x=1:3*1.2-0.5,y=sl_stat+1980,paste0(sl_stat,"\n",100*round(as.numeric(sl_stat/sum(sl_stat)),4),"%"),cex = 0.75)
``` 

#### RidgePlot    
##### with ridge plots, raw    
```{r fig.width=12,fig.height=6,message=FALSE, warning=FALSE}
FB.seur@active.ident <- factor(FB.seur$HTO_maxID,levels=rownames(FB.seur))

RidgePlot(FB.seur, assay = 'HTO', features = rownames(FB.seur[['HTO']]),same.y.lims= TRUE, ncol=4,
          cols = color.FB) 
```


##### with ridge plots, filtered    

```{r fig.width=12,fig.height=7,message=FALSE, warning=FALSE}
FB.seur@meta.data$hash.ID.sort <- factor(as.character(FB.seur@meta.data$hash.ID),levels = c("Doublet","Negative",levels(FB.seur@active.ident)))
RidgePlot(FB.seur, assay = 'HTO', features = rownames(FB.seur[['HTO']]), ncol=4,same.y.lims= TRUE,
          cols = c("#FF6C91","lightgrey",color.FB),group.by = "hash.ID.sort") 
```

#### tSNE for HTOs     
     
     
```{r message=FALSE, warning=FALSE}
# first, remove negative cells from th object
#FB.seur.subset <- FB.seur

# Calculate a tSNE embedding of the HTO data
#DefaultAssay(FB.seur.subset) <- "HTO"

#FB.seur.subset <- ScaleData(FB.seur.subset, features = rownames(FB.seur.subset), verbose = FALSE)
#FB.seur.subset <- RunPCA(FB.seur.subset, features = rownames(FB.seur.subset), approx = FALSE)
#FB.seur.subset <- RunTSNE(FB.seur.subset, dims = 1:8, perplexity = 500, check_duplicates = FALSE)


#saveRDS(FB.seur.subset, "FB0809.seur.subset.rds")
FB.seur.subset <- readRDS("FB0809.seur.subset.rds")
```


```{r fig.width=5.5,fig.height=8.6}
multiplot(
DimPlot(FB.seur.subset, order = rev(levels(FB.seur@active.ident)),
        cols = color.FB) + labs(title = "FB Max_ID"), 
DimPlot(FB.seur.subset, order = c("Doublet",rev(levels(FB.seur@active.ident)),"Negative"), group.by = 'hash.ID.sort', label = F,
        cols = c("lightgrey",color.FB,"#FF6C91")) + 
  labs(title = "FB Filtered_ID") + theme(plot.title = element_text(hjust = 0)) ,
cols = 1)
```


this is not bad ! with comfortable colors/shapes.      
(while for sn10x data, might should be careful to open your eyes~)       
          

### stat    

```{r}
FB.seur$HTO_classification.global <- factor(as.character(FB.seur$HTO_classification.global),
                                            levels = c("Doublet","Negative","Singlet"))
Idents(FB.seur) <- "HTO_classification.global"
VlnPlot(FB.seur, features = "nCount_RNA", pt.size = 0, log = TRUE, cols = c("#F8766D","lightgrey","#00BA38"))
VlnPlot(FB.seur, features = "nCount_RNA", pt.size = 0, log = TRUE, cols = c("#F8766D","lightgrey","#00BA38")) + geom_jitter(alpha=0.015)
```

```{r}
table(FB.seur@meta.data$hash.ID.sort)
```

```{r fig.width=7.2,fig.height=4}
par(mar=c(6,4,2,3))
sl_stat <- table(FB.seur$hash.ID.sort)
barplot(sl_stat,ylim = c(0,12800),
        col = c("#FF6C91","lightgrey",color.FB),
        main = "Feature Barcode statistics",cex.names = 0.75, xaxt = "n")
axis(1,1:10*1.2-0.48 ,levels(FB.seur@meta.data$hash.ID.sort), las=3, cex.axis=0.85)
text(x=1:10*1.2-0.45,y=sl_stat+875,paste0(sl_stat,"\n",100*round(as.numeric(sl_stat/sum(sl_stat)),4),"%"),cex = 0.75)
```


## GEX    
    
mainly follow https://satijalab.org/seurat/articles/pbmc3k_tutorial.html    


```{r message=FALSE, warning=FALSE}
GEX.seur <- CreateSeuratObject(counts = GEX,
                               min.cells = 3,
                               min.features = 200,
                               project = "ILC3_GEX")
GEX.seur
```


### filtering    

#### MT genes   
```{r}
grep("^mt-",rownames(GEX),value = T)
```

```{r}
MT_gene <- grep("^mt-",rownames(GEX.seur),value = T)
GEX.seur[["percent.mt"]] <- PercentageFeatureSet(GEX.seur, features = MT_gene)

RB_gene <- grep("^Rps|^Rpl",rownames(GEX.seur),value = T)
GEX.seur[["percent.rb"]] <- PercentageFeatureSet(GEX.seur, features = RB_gene)

# Visualize QC metrics as a violin plot
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0.01)
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0)
```

```{r fig.width=15,fig.height=4}
plota <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plotb <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
plotc <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.rb")
plota + plotb + plotc
```

#### add FB info      

```{r}
GEX.seur$FB.info <- FB.seur@meta.data[rownames(GEX.seur@meta.data),"hash.ID"]
#head(GEX.seur@meta.data)
```


```{r}
table(GEX.seur$FB.info)
```


```{r fig.width=8.1, fig.height=8}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, split.by = "FB.info")
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0, split.by = "FB.info")
```


```{r}
GEX.seur <- subset(GEX.seur, subset = FB.info!="Doublet" & FB.info!="Negative")
GEX.seur
```

```{r fig.height=5, fig.width=8, message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0.1)
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0)
```

```{r fig.height=8, fig.width=8.1, message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, split.by = "FB.info")
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0, split.by = "FB.info")
```


```{r fig.width=15,fig.height=4}
plota <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plotb <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
plotc <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.rb") 
plota + plotb + plotc
```

#### filtering        


```{r}
GEX.seur <- subset(GEX.seur, subset = percent.mt < 7.5 & nFeature_RNA > 500 & nFeature_RNA < 3000 & nCount_RNA < 10000)
GEX.seur
```

filtered obj        
```{r fig.height=5, fig.width=8, message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0.1)
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0)
```

```{r fig.height=8, fig.width=8.1, message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = color.FBraw,
        pt.size = 0.1, split.by = "FB.info")
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        cols = color.FBraw,
        pt.size = 0, split.by = "FB.info")
```

```{r fig.width=15,fig.height=4}
plota <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plotb <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
plotc <- FeatureScatter(GEX.seur, feature1 = "nCount_RNA", feature2 = "percent.rb") 
plota + plotb + plotc
```

```{r fig.width=7.2,fig.height=4}
par(mar=c(6,4,2,3))
sl_stat <- table(FB.seur$hash.ID.sort)
barplot(sl_stat,ylim = c(0,12800),
        col = c("#FF6C91","lightgrey",color.FB),
        main = "Feature Barcode statistics",cex.names = 0.75, xaxt = "n")
axis(1,1:10*1.2-0.48 ,levels(FB.seur@meta.data$hash.ID.sort), las=3, cex.axis=0.85)
text(x=1:10*1.2-0.45,y=sl_stat+875,paste0(sl_stat,"\n",100*round(as.numeric(sl_stat/sum(sl_stat)),4),"%"),cex = 0.75)
```

```{r fig.width=7.2,fig.height=4}
par(mar=c(6,4,2,3))
sl_stat <- table(GEX.seur$FB.info)
barplot(sl_stat,ylim = c(0,4500),
        col = c("#FF6C91","lightgrey",color.FB),
        main = "Feature Barcode statistics",cex.names = 0.75, xaxt = "n")
axis(1,1:10*1.2-0.48 ,levels(FB.seur@meta.data$hash.ID.sort), las=3, cex.axis=0.85)
text(x=1:10*1.2-0.45,y=sl_stat+325,paste0(sl_stat,"\n",100*round(as.numeric(sl_stat/sum(sl_stat)),4),"%"),cex = 0.75)
```

### Markers and Clusters            

#### Normalizing          

```{r}
GEX.seur <- NormalizeData(GEX.seur, normalization.method = "LogNormalize", scale.factor = 10000)
```

#### check cellcycle       
```{r}
s.genes <- Hmisc::capitalize(tolower(cc.genes.updated.2019$s.genes))
g2m.genes <- Hmisc::capitalize(tolower(cc.genes.updated.2019$g2m.genes))


GEX.seur <- CellCycleScoring(GEX.seur,
                             s.features = s.genes,
                             g2m.features = g2m.genes)
```


```{r message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("S.Score", "G2M.Score"), group.by = "FB.info", cols = color.FB,
    ncol = 2, pt.size = 0.0)  & geom_jitter(alpha=0.012)
```

```{r message=FALSE, warning=FALSE}
GEX.seur.cc <- GEX.seur

GEX.seur.cc <- NormalizeData(GEX.seur.cc)
GEX.seur.cc <- FindVariableFeatures(GEX.seur.cc)
GEX.seur.cc <- ScaleData(GEX.seur.cc)
Idents(GEX.seur.cc) <- "Phase"
RidgePlot(GEX.seur.cc, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol = 2)
```

```{r message=FALSE, warning=FALSE}
GEX.seur.cc <- RunPCA(GEX.seur.cc, features = c(s.genes, g2m.genes))
DimPlot(GEX.seur.cc, reduction = 'pca')
```

some few cycling should exist.           


#### VariableFeatures          


```{r message=FALSE, warning=FALSE,fig.width=12,fig.height=5}
GEX.seur <- FindVariableFeatures(GEX.seur, selection.method = "vst", nfeatures = 1500)

# Identify the 10 most highly variable genes
#top10 <- head(VariableFeatures(GEX.seur), 10)
top20 <- head(VariableFeatures(GEX.seur), 20)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(GEX.seur)
plot2 <- LabelPoints(plot = plot1, points = top20, repel = T)
plot1 + plot2
```

```{r}
head(VariableFeatures(GEX.seur), 200)
```

```{r}
GEX.seur <- ScaleData(GEX.seur, features = rownames(GEX.seur))
```


#### PCA       

```{r}
# exclude MT genes  and more 

#DIG <- grep("^Tra|^Trb|^Trg|^Trd|^Tcr|^Igm|^Igh|^Igk|^Igl|Jchain|^Hsp|^Rps|^Rpl|Hbb-|Hba-|^Dnaj|^AY|^Gm|^Hist",rownames(GEX.seur),value = T)

DIG <- grep("^Tra|^Trb|^Trg|^Trd|^Tcr|^Igm|^Igh|^Igk|^Igl|Jchain|Mzb1|Vpreb|Lars2|Jun|Fos|^Hsp|^Rps|^Rpl|Hbb-|Hba-|^Dnaj|^AY|^AC|^AA|^AW|^BC|^Gm|^Hist|Rik$|-ps",
            rownames(GEX.seur),value = T)
CC_gene <- Hmisc::capitalize(tolower(as.vector(unlist(cc.genes.updated.2019))))


VariableFeatures(GEX.seur) <- setdiff(VariableFeatures(object = GEX.seur),
                                                c(MT_gene,
                                                  DIG, 
                                                  CC_gene) )

GEX.seur <- RunPCA(GEX.seur, features = VariableFeatures(GEX.seur))
```


```{r}
length(setdiff(VariableFeatures(object = GEX.seur),
                                                c(MT_gene,DIG,CC_gene)))
setdiff(VariableFeatures(object = GEX.seur),
                                                c(MT_gene,DIG,CC_gene))[1:300]
```

```{r fig.width=11.8,fig.height=4.5}
DimPlot(GEX.seur, reduction = "pca",dims = 1:2, group.by = "orig.ident") +
  DimPlot(GEX.seur, reduction = "pca",dims = 3:4, group.by = "orig.ident")
```

```{r fig.width=11.5,fig.height=4.5}
DimPlot(GEX.seur, reduction = "pca",dims = 1:2, group.by = "FB.info", cols = color.FB) +
  DimPlot(GEX.seur, reduction = "pca",dims = 3:4, group.by = "FB.info", cols = color.FB)
```

```{r pcsheat,fig.width=12,fig.height=15}
DimHeatmap(GEX.seur, dims = 1:12, cells = 1500, balanced = TRUE,ncol = 4)
```

##### decide PCs to use          
     
```{r}
ElbowPlot(GEX.seur,ndims = 50)
```

```{r}
PCs <- 1:27
GEX.seur <- FindNeighbors(GEX.seur, dims = PCs, k.param = 10)
GEX.seur <- FindClusters(GEX.seur, method = 'igraph' ,resolution = 2.5)
```

#### Run UMAP/tSNE    

```{r}
GEX.seur <- RunTSNE(GEX.seur, dims=PCs, complexity = 100)
GEX.seur <- RunUMAP(GEX.seur, dims=PCs, n.neighbors = 20, seed.use = 153)
```


```{r fig.width=11.5,fig.height=4.5}
DimPlot(GEX.seur, reduction = "tsne", label = T) + DimPlot(GEX.seur, reduction = "umap", label = T)
```

```{r fig.width=7.5,fig.height=6}
FeaturePlot(GEX.seur, reduction = "umap", features = c("nFeature_RNA","nCount_RNA","percent.mt","percent.rb"))
```

```{r  fig.width=12.5, fig.height=6.4}
DimPlot(GEX.seur, label = F, pt.size = 0.05, repel = F, reduction = 'tsne', group.by = "FB.info", split.by = "FB.info", 
        ncol = 4, cols = color.FB)
DimPlot(GEX.seur, label = F, pt.size = 0.05, repel = F, reduction = 'umap', group.by = "FB.info", split.by = "FB.info", 
        ncol = 4, cols = color.FB)
``` 

```{r}
GEX.seur$cnt <- factor(gsub(".1$|.2$|.3$|.4$","",as.character(GEX.seur$FB.info)),
                       levels = c("CTL",
                                  "CKO"))
color.cnt <- color.FB[c(5,1)]
```


```{r fig.width=10.5,fig.height=4.5}
DimPlot(GEX.seur, reduction = "umap", label = T, group.by = "seurat_clusters") +
  DimPlot(GEX.seur, reduction = "umap", label = F, group.by = "cnt", cols = color.cnt)
```


```{r eval=FALSE, fig.height=4.5, fig.width=10.5, include=FALSE}
umap.cluster <- list()

for(nn in 0:37){
  
  xx=as.character(nn)
  umap.cluster[[xx]] <- (DimPlot(subset(GEX.seur, subset = seurat_clusters %in% c(xx)), cols = scales::hue_pal()(38)[nn+1],
                                reduction = "umap", label = T, group.by = "seurat_clusters") + xlim(-7,10) + ylim(-8,11.5) ) +
                          (DimPlot(subset(GEX.seur, subset = seurat_clusters %in% c(xx)), 
                                  reduction = "umap", label = F, group.by = "cnt", cols = color.cnt) + xlim(-7,10) + ylim(-8,11.5) )
  
}

#umap.cluster

```




#### check some markers  

```{r}
markers.mig <- c("Ptprc","Cd3d","Cd3e","Trdc",
                 "Tbx21","Ifng","Gata3","Il5",
                 "Il10","Il4","Calca","Hilpda",
                 "Rorc","Fcer1g","Il22","Il17a",
                 "Cd4","Cd8a")
```


```{r fig.width=12, fig.height=12.5}
FeaturePlot(GEX.seur, 
            features = markers.mig,
            ncol = 4)
```

```{r fig.width=3, fig.height=2.5}
FeaturePlot(GEX.seur, features = "Rorc")
```


```{r fig.width=3, fig.height=2.5}
FeaturePlot(GEX.seur, features = "Il22")
```


```{r fig.width=6, fig.height=7.5}
DotPlot(GEX.seur, features = markers.mig, group.by = "seurat_clusters",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1))+ scale_y_discrete(limits=rev)
```


```{r fig.width=6, fig.height=5}
DotPlot(GEX.seur, features = markers.mig, group.by = "FB.info",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1))+ scale_y_discrete(limits=rev)
```


```{r}
# well, this list of markers is from result of previous test analysis
markers.sort <- c("Cd3d","Cd3e","Ccr7","Lef1",
                     "S1pr1","Sell","Sox4","Tcrg-C2",
                     "Izumo1r","Cd4","Il10","Hopx",
                     "Foxp3","Maf","Nebl","Il17a",
                     "Top2a","Mki67","Pcna","Mcm6",
                     "Ly6a","Gzma","Cd40lg","Ifng",
                     "Ccl5","Klrd1","Cd7","Itgae",
                     "Ahr",
                     "Igkc","Igha","Iglc1","Jchain",
                     "Trdc","Trdv4","S100a6","Cd163l1",
                     "Pdcd1",
                     "Jun","Fos","Egr1","Hmgn3",
                     "Rorc","Ccr6","Cd74","Cd83",
                     "Cd81","Il17f","Prr29","Car2",
                     "Tmem176a","Tmem176b","Gzmc","Irf8",
                     "Cxcl2","Cxcl3","Gzmb","Ncr1",
                     "Cxcr6","Upp1","Klrb1c","Cd69",
                     "Apoe","Lyz2","H2-Aa","C1qb",
                     "Fcer1g",
                     "Ifit1","Isg20","Stat1","Cxcl10",
                     "Calca","Gata3","Il17rb","Il13",
                     "Il4","Hilpda","Ar","Il1rl1",
                     "Il5","Cd24a","Ccl1","Areg"
                     )
```


```{r fig.width=16.5, fig.height=7}
#DotPlot(GEX.seur, features = markers.sort, group.by = "sort_clusters",
DotPlot(GEX.seur, features = markers.sort, group.by = "seurat_clusters",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1))+ scale_y_discrete(limits=rev)

``` 



#### DoubletFinder       

```{r}
library(DoubletFinder)
```

```{r include=FALSE}
# to find a better pk
sweep.res.list <- paramSweep_v3(GEX.seur, PCs = PCs, sct = FALSE) 
```

```{r message=FALSE, warning=FALSE}
for(i in 1:length(sweep.res.list)){
  if(length(sweep.res.list[[i]]$pANN[is.nan(sweep.res.list[[i]]$pANN)]) !=0){
    if(i!=1){
      sweep.res.list[[i]] <- sweep.res.list[[i-1]]
    }else(
      sweep.res.list[[i]] <- sweep.res.list[[i+1]]
    )
  }
}
sweep.stats <- summarizeSweep(sweep.res.list, GT=FALSE)
bcmvn <- find.pK(sweep.stats)
```

```{r}
pk_v <- as.numeric(as.character(bcmvn$pK))
pk_good <- pk_v[bcmvn$BCmetric==max(bcmvn$BCmetric)]
```

```{r}
# specify expected doublet number     
nExp_poi <- round(0.05*length(colnames(GEX.seur)))

GEX.seur <- doubletFinder_v3(GEX.seur, PCs = PCs, pN = 0.25, pK = pk_good, 
                             nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)
colnames(GEX.seur@meta.data)[ncol(GEX.seur@meta.data)] <- "DoubletFinder0.05"
```

```{r message=FALSE, warning=FALSE}
# specify expected doublet number     
nExp_poi <- round(0.1*length(colnames(GEX.seur)))

GEX.seur <- doubletFinder_v3(GEX.seur, PCs = PCs, pN = 0.25, pK = pk_good, 
                             nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)
colnames(GEX.seur@meta.data)[ncol(GEX.seur@meta.data)] <- "DoubletFinder0.1"
```


```{r fig.width=10.8,fig.height=4.5}
DimPlot(GEX.seur, reduction = "umap", group.by = "DoubletFinder0.05") +
  DimPlot(GEX.seur, reduction = "umap", group.by = "DoubletFinder0.1")
```


```{r fig.width=10.5,fig.height=4.5}
DimPlot(GEX.seur, reduction = "umap", label = T, group.by = "seurat_clusters") +
  DimPlot(GEX.seur, reduction = "umap", group.by = "DoubletFinder0.1")
```


```{r eval=FALSE, fig.height=4.5, fig.width=10.5, include=FALSE}
umap.cluster.DF <- list()

for(nn in 0:37){
  
  xx=as.character(nn)
  umap.cluster.DF[[xx]] <- (DimPlot(subset(GEX.seur, subset = seurat_clusters %in% c(xx)), cols = scales::hue_pal()(38)[nn+1],
                                reduction = "umap", label = T, group.by = "seurat_clusters") + xlim(-7,10) + ylim(-8,11.5) ) +
                          (DimPlot(subset(GEX.seur, subset = seurat_clusters %in% c(xx)), 
                                  reduction = "umap", label = F, group.by = "DoubletFinder0.1") + xlim(-7,10) + ylim(-8,11.5) )
  
}

#umap.cluster.DF

```


#### sort_clusters          


```{r}
GEX.seur$sort_clusters <- factor(as.character(GEX.seur$seurat_clusters),
                                 levels = c(2,11,    # naive  
                                            32, 29,  # 
                                            24,
                                            18, 13, 27, 
                                            16, 31, 33, 9, 20,  
                                            
                                            23,
                                            
                                            12, 22,5,10,6, 35, 0,  
                                            17,4,1,3,15, 19, 14, 25, 36, 26, 
                                            7, 21, 
                                            
                                            8,28, 30,37,34))
```




```{r}
# previous ref markers
markers.sort <- c("Cd3d","Cd3e","Ccr7","Lef1",
                     "S1pr1","Sell","Sox4","Tcrg-C2",
                     "Izumo1r","Cd4","Il10","Hopx",    "Cd8a","Cd8b1",
                     "Foxp3","Maf","Nebl","Il17a",
                     "Top2a","Mki67","Pcna","Mcm6",
                     "Ly6a","Gzma","Cd40lg","Ifng",
                     "Ccl5","Klrd1","Cd7","Itgae",
                     "Ahr",
                     "Igkc","Igha","Iglc1","Jchain",
                     "Trdc","Trdv4","S100a6","Cd163l1",
                     "Pdcd1",
                     "Jun","Fos","Egr1","Hmgn3",
                     "Rorc","Ccr6","Cd74","Cd83",
                     "Cd81","Il17f","Prr29","Car2",
                     "Tmem176a","Tmem176b","Gzmc","Irf8",
                     "Cxcl2","Cxcl3","Gzmb","Ncr1",
                     "Cxcr6","Upp1","Klrb1c","Cd69",
                     "Apoe","Lyz2","H2-Aa","C1qb",
                     "Fcer1g",
                     "Ifit1","Isg20","Stat1","Cxcl10",
                     "Calca","Gata3","Il17rb","Il13",
                     "Il4","Hilpda","Ar","Il1rl1",
                     "Il5","Cd24a","Ccl1","Areg"
                     )
```


```{r fig.width=16.5, fig.height=7}
DotPlot(GEX.seur, features = markers.sort, group.by = "sort_clusters",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1))+ scale_y_discrete(limits=rev)

``` 

```{r fig.width=14.1, fig.height=6.}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "sort_clusters")
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0, group.by = "sort_clusters")
```

```{r fig.width=10.1, fig.height=4.}
VlnPlot(GEX.seur, features = c("percent.rb"), 
        #ncol = 3, 
        ncol = 1,
        cols = color.cnt,
        pt.size = 0.1, group.by = "sort_clusters", split.by = "cnt")
VlnPlot(GEX.seur, features = c("percent.rb"), 
        #ncol = 3, 
        ncol = 1,
        cols = color.cnt,
        pt.size = 0, group.by = "sort_clusters", split.by = "cnt")
```

```{r}
table(GEX.seur@meta.data[,c("sort_clusters","DoubletFinder0.05")])
```

```{r}
table(GEX.seur@meta.data[,c("sort_clusters","DoubletFinder0.1")])
```


### markers       

```{r message=FALSE, warning=FALSE, paged.print=TRUE}
# find markers for every cluster compared to all remaining cells, report only the positive ones
Idents(GEX.seur) <- "sort_clusters"

#GEX.markers.pre <- FindAllMarkers(GEX.seur, only.pos = TRUE, min.pct = 0.05,
#                                  test.use = "MAST",
#                                  #test.use = "wilcox",
#                                  logfc.threshold = 0.25)
GEX.markers.pre <- read.table("20230927_10x_SZJ.new20240809.sort_markers.csv", header = TRUE, sep = ",")
#GEX.markers.pre %>% group_by(cluster) %>% top_n(n = 8, wt = avg_log2FC)
```



```{r eval=FALSE, include=FALSE}
write.table(GEX.markers.pre, 
            "20230927_10x_SZJ.new20240809.sort_markers.csv", 
            col.names = TRUE,
            row.names = FALSE,
            quote = F,
            sep = ",")
```



```{r}
GEX.markers.pre$cluster <- factor(as.character(GEX.markers.pre$cluster),
                          levels = levels(GEX.seur$sort_clusters))



markers.pre_t32 <- (GEX.markers.pre %>% group_by(cluster) %>% 
                  filter(pct.1>0.1) %>%
                   top_n(n = 32, wt = avg_log2FC) %>%
                   ungroup() %>%
  arrange(desc(avg_log2FC*pct.1),gene) %>%
                             distinct(gene, .keep_all = TRUE) %>%
                             arrange(cluster,p_val_adj))$gene


markers.pre_t48 <- (GEX.markers.pre %>% group_by(cluster) %>% 
                  filter(pct.1>0.05 & gene %in% grep("Rps|Rpl|mt-",GEX.markers.pre$gene,invert = T,value = T)) %>%
                   top_n(n = 48, wt = avg_log2FC) %>%
                   ungroup() %>%
  arrange(desc(avg_log2FC*pct.1),gene) %>%
                             distinct(gene, .keep_all = TRUE) %>%
                             arrange(cluster,p_val_adj))$gene

markers.pre_t60 <- (GEX.markers.pre %>% group_by(cluster) %>% 
                  filter(pct.1>0.01 & gene %in% grep("Rps|Rpl|mt-",GEX.markers.pre$gene,invert = T,value = T)) %>%
                   top_n(n = 60, wt = avg_log2FC) %>%
                    filter(p_val_adj < 0.01) %>%
                   ungroup() %>%
  arrange(desc(avg_log2FC*pct.1),gene) %>%
                             distinct(gene, .keep_all = TRUE) %>%
                             arrange(cluster,p_val_adj))$gene

markers.pre_t120 <- (GEX.markers.pre %>% group_by(cluster) %>% 
                  filter(pct.1>0.01 & gene %in% grep("Rps|Rpl|mt-",GEX.markers.pre$gene,invert = T,value = T)) %>%
                   top_n(n = 120, wt = avg_log2FC) %>%
                    filter(p_val_adj < 0.01) %>%
                   ungroup() %>%
  arrange(desc(avg_log2FC*pct.1),gene) %>%
                             distinct(gene, .keep_all = TRUE) %>%
                             arrange(cluster,p_val_adj))$gene
```


```{r}
861/60
861/64
861/65
```

```{r fig.width=9.6, fig.height=11.4}

pp.t60 <- list()

for(i in 1:14){
pp.t60[[i]] <- DotPlot(GEX.seur, features = rev(markers.pre_t60[(64*i-63):(64*i)]))  + coord_flip() + 
           theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9.6))
}

pp.t60
```


```{r}
nnn = 1447
nnn/60
nnn/64
nnn/65
```

```{r fig.width=9.6, fig.height=11.4}

pp.t120 <- list()

for(i in 1:23){
pp.t120[[i]] <- DotPlot(GEX.seur, features = rev(markers.pre_t120[(64*i-63):(64*i)]))  + coord_flip() + 
           theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9.6))
}

#pp.t120
```





## Cell Composition              


```{r fig.width=10.2, fig.height=5.4}
cowplot::plot_grid(
pheatmap::pheatmap(table(FB.info=GEX.seur$FB.info,
      clusters=GEX.seur$sort_clusters)[c(3:10),],
                   main = "Cell Count",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.0f", legend = F, silent = T)$gtable,

pheatmap::pheatmap(100*rowRatio(table(FB.info=GEX.seur$FB.info,
                                clusters=GEX.seur$sort_clusters)[c(3:10),]),
                   main = "Cell Ratio",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.1f", legend = F, silent =T)$gtable,
ncol = 1)
```


## check            

```{r fig.height=15, fig.width=15.1, message=FALSE, warning=FALSE}
VlnPlot(GEX.seur, features = c("Cd3d","Cd4","Rorc","Il17a",
                               "Il22","Il17f","H2-K1","H2-Q7",
                               "Rps19","Tnfsf11","Ahr","Gzmc"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "sort_clusters") + NoLegend()

```



```{r fig.height=12, fig.width=15}
VlnPlot(GEX.seur, features = c("Gzmk","Gzma",
                               "Tbx21","Nkg7",
                               "Il4","Xcl1",
                               "Klrb1c","Cd160"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "sort_clusters") + NoLegend()
```



## preAnno             


```{r}
####
##
#Cycling:  C26 - ILC3.Ncr1
#          C27 - Th17 (most)
#
#Mix:    C29/33/35
#        C23 - Bcell
#        C36 - Myeloid
#
#Tcell
#  Tn(Naive)    C2/11   Sell+Ccr7+
#       C32  Cd4-Trac-,very few Cd8a+,Pdcd1+   next to remove it
#  Izumo1r+Lrig1+,very few Foxp3+   C24 (like a transit to Treg)
#      let it be Treg.0
#  Treg    C18   Foxp3+ (Lzumo1r low) Il10+
#  Th17    C13   Il17a+
#      Cd40lg (hi) Ifng (hi) Fos/Egr1 (hi)   C31
#  Th1   C16  Gzmk+ Gzma+
#
#  innate like (previous  Tin )
#    iNKT   C9   Cd4-Tbx21+Il4+
#    MAIT   C20    Cd4-Cd160+Xcl1+
#
#  Th2-like   C34  Cd3d+Cd4+Foxp3-Gata3+
#  gdT       C7/21   Trdc+Trdv4+Tcrg-C1+
#                    Cd163l1+Blk+Mmp25+
#
#ILC2       C8/28/30      
#      C37   Ifit1+Isg20+Stat1+
#
#ILC3
#  Ccr6+      C12/22/5/10/6
#             C22  Cd74+MHCII+(H2-Aa/H2-Eb1-H2-Ab1)
#             C12  Jun (hi) Fos (hi) Rorc (hi)
#
#  Ccr6-Ncr1-    C0
#
#  Ncr1+      C17/4/1/3/15/19/14/25
#             C14  Jun (hi) Fos (hi) Rorc (hi)
#             C19   Cxcl9+Serpina3f+Irgm1+ Tgtp1/2+
#
##
####
```




```{r}
GEX.seur$preAnno <- as.character(GEX.seur$seurat_clusters)


GEX.seur$preAnno[GEX.seur$preAnno %in% c(2,11)] <- "Tn"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(24,18)] <- "Treg"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(13)] <- "Th17"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(27)] <- "Th.CC"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(16)] <- "Th1"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(31)] <- "Th.Egr1"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(9)] <- "iNKT"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(20)] <- "MAIT"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(34)] <- "Th2"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(7,21)] <- "gdT"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(8,28,30)] <- "ILC2"
GEX.seur$preAnno[GEX.seur$preAnno %in% c(37)] <- "ILC2.Ifit1"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(26)] <- "ILC3.CC"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(17,4,1,3,15,19,14,25)] <- "ILC3.Ncr1"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(0)] <- "ILC3.nn"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(12,22,5,10,6)] <- "ILC3.Ccr6"

GEX.seur$preAnno[GEX.seur$preAnno %in% c(29,33,35,23,36,32)] <- "Mix"

GEX.seur$preAnno <- factor(GEX.seur$preAnno,
                           levels = c("Tn","Treg",
                                      "Th17","Th.CC","Th1","Th.Egr1","Th2",
                                      "iNKT","MAIT","gdT",
                                      "ILC2","ILC2.Ifit1",
                                      "ILC3.CC","ILC3.Ncr1","ILC3.nn","ILC3.Ccr6",
                                      "Mix"))


```



```{r fig.width=11.5,fig.height=4.5}
DimPlot(GEX.seur, reduction = "umap", label = T, group.by = "seurat_clusters") +
  DimPlot(GEX.seur, reduction = "umap", label = T, label.size = 3.2, group.by = "preAnno")
```


```{r fig.width=10.2, fig.height=5.4}
cowplot::plot_grid(
pheatmap::pheatmap(table(FB.info=GEX.seur$FB.info,
      clusters=GEX.seur$preAnno)[c(3:10),],
                   main = "Cell Count",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.0f", legend = F, silent = T)$gtable,

pheatmap::pheatmap(100*rowRatio(table(FB.info=GEX.seur$FB.info,
                                clusters=GEX.seur$preAnno)[c(3:10),]),
                   main = "Cell Ratio",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.1f", legend = F, silent =T)$gtable,
ncol = 1)
```


### clean up                 

```{r fig.width=12.1, fig.height=6.}
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "preAnno")
VlnPlot(GEX.seur, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0, group.by = "preAnno")
```


```{r}
GEX.pure <- subset(GEX.seur, subset = preAnno != "Mix")
GEX.pure <- subset(GEX.pure, subset = DoubletFinder0.1=="Singlet" | preAnno %in% c("Th.CC","ILC3.CC"))
#GEX.pure <- subset(GEX.pure, subset = percent.mt < 7.5 & nFeature_RNA > 500 & nFeature_RNA < 3000 & nCount_RNA < 10000)
GEX.pure
```

```{r fig.width=11.5,fig.height=4.5}
DimPlot(GEX.pure, reduction = "umap", label = T, group.by = "seurat_clusters") +
  DimPlot(GEX.pure, reduction = "umap", label = T, label.size = 3.2, group.by = "preAnno")
```


```{r fig.width=12.1, fig.height=6.}
VlnPlot(GEX.pure, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "preAnno")
VlnPlot(GEX.pure, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0, group.by = "preAnno")
```


```{r fig.width=8.2, fig.height=5.4}
cowplot::plot_grid(
pheatmap::pheatmap(table(FB.info=GEX.pure$FB.info,
      clusters=GEX.pure$preAnno)[c(3:10),],
                   main = "Cell Count",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.0f", legend = F, silent = T)$gtable,

pheatmap::pheatmap(100*rowRatio(table(FB.info=GEX.pure$FB.info,
                                clusters=GEX.pure$preAnno)[c(3:10),]),
                   main = "Cell Ratio",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.1f", legend = F, silent =T)$gtable,
ncol = 1)
```


```{r}
# previous ref markers
markers.sort <- c("Cd3d","Cd3e","Ccr7","Lef1",
                     "S1pr1","Sell","Sox4","Tcrg-C2",
                     "Izumo1r","Cd4","Il10","Hopx",    "Cd8a","Cd8b1",
                     "Foxp3","Maf","Nebl","Il17a",
                     "Top2a","Mki67","Pcna","Mcm6",
                     "Ly6a","Gzma","Cd40lg","Ifng",
                     "Ccl5","Klrd1","Cd7","Itgae",
                     "Ahr",
                     "Igkc","Igha","Iglc1","Jchain",
                     "Trdc","Trdv4","S100a6","Cd163l1",
                     "Pdcd1",
                     "Jun","Fos","Egr1","Hmgn3",
                     "Rorc","Ccr6","Cd74","Cd83",
                     "Cd81","Il17f","Prr29","Car2",
                     "Tmem176a","Tmem176b","Gzmc","Irf8",
                     "Cxcl2","Cxcl3","Gzmb","Ncr1",
                     "Cxcr6","Upp1","Klrb1c","Cd69",
                     "Apoe","Lyz2","H2-Aa","C1qb",
                     "Fcer1g",
                     "Ifit1","Isg20","Stat1","Cxcl10",
                     "Calca","Gata3","Il17rb","Il13",
                     "Il4","Hilpda","Ar","Il1rl1",
                     "Il5","Cd24a","Ccl1","Areg"
                     )
```



```{r fig.width=18.5, fig.height=6}
markers.new <-     c("Cd3d","Cd3e","Cd4","Cd8a",
                     "Klf2","Ccr7","Lef1","S1pr1",
                     "Tubb2a","Rflnb","Igfbp4","Sell",
                     "Itm2a","Itga6","Smc4",#"Sox4","Pdcd1",
                     "Ptprc",#"Fam241a","Slamf6",
                     "Izumo1r","Tbc1d4",
                     "Ctla4","Il10", 
                     "Hopx","Foxp3","Ccr2","Maf",
                     "Capg","Slamf1","Icos",
                     "Il17a","Nebl","S100a11","S100a10",
                     "Top2a","Mki67","Pcna","Mcm6",
                     
                     "Fasl","Gzma","Gzmk","Ccr5",
                     "Il12rb2",
                     "Jun","Fos","Ier2","Egr1",
                     
                     "Ifng", "Tnf",
                     "Tbx21",
                     "Klrb1c","Xcl1","Nkg7","Plac8","Cd160",
                     
                     
                     
                     "Ccl5",
                     "Klrd1","Cd7","Itgae",
                     
                     "Tcrg-C1","Trdv4","Cd163l1","Blk","Mmp25",
                     
                     "Calca","Gata3","Il17rb","Il13",
                     "Csf2","Dgat2",
                     "Il4","Hilpda","Ar","Il1rl1",
                     "Il5",#"Cd24a",
                     "Ccl1","Areg","Ahr",
                     "Cxcl10",
                     "Gbp6","Stat1","Isg15","Ifit1",
                     
                     "Ncr1","Gzmc","Gzmb","Irf8",
                     "Cxcr6","Irf7","Slc16a6","Klrk1",
                     "Tmem176b","Fcer1g","Car2","Ckb",
                     "Cd69","Jaml","Cxcl9",
                     "Il22","Gem","Pxdc1","Cdc14a",
                     "Sdc4","Vegfa","Bst2","Nmrk1",
                     "Hmgn3","Cd81","Cx3cl1",
                     "Cd74","H2-Aa","H2-Eb1","H2-Ab1",
                     
                     
                     "Ccr6","Batf3","Rorc","Maff",
                     "Icam1","Il18r1",
                     
                     "Il17f"
                     
                     )

DotPlot(GEX.pure, features = markers.new, group.by = "preAnno",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1, size = 9)) + 
  scale_y_discrete(limits=rev)
``` 

```{r fig.width=18.5, fig.height=6}

DotPlot(GEX.seur, features = markers.new, group.by = "preAnno",
        cols = c("midnightblue","darkorange1")) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1, size = 9)) + 
  scale_y_discrete(limits=rev)
``` 

```{r}
#saveRDS(GEX.seur, "./20230927_10x_SZJ.new20270809.preAnno.rds")
```


```{r fig.height=15, fig.width=12.1, message=FALSE, warning=FALSE}
VlnPlot(GEX.pure, features = c("Cd3d","Cd4","Rorc","Il17a",
                               "Tbx21","Xcl1","Cd160","Cd7",
                               "Klrb1c","Ifng","Mr1","Cd1d1"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "preAnno") + NoLegend()

```


```{r fig.height=15, fig.width=12.1, message=FALSE, warning=FALSE}
VlnPlot(GEX.pure, features = c("Cd160","Plac8","Klrd1","Itgae",
                               "Tcrg-C1","Trdv4","Cd163l1","Blk",
                               "Gzma","Gzmk","Ccr5","Ccl5"), 
        #ncol = 3, 
        ncol = 2,
        #cols = c("#FF6C91","lightgrey",color.FB),
        pt.size = 0.1, group.by = "preAnno") + NoLegend()

```




#### stat           

```{r}
GEX.pure$rep <- gsub("CKO.|CTL.","rep",as.character(GEX.pure$FB.info))
GEX.pure$rep[1:5]
```

```{r}
GEX.pure$preAnno <- factor(as.character(GEX.pure$preAnno),
                           levels = setdiff(levels(GEX.seur$preAnno),"Mix"))
```



```{r}
stat_preAnno <- GEX.pure@meta.data[,c("cnt","rep","preAnno")]

stat_preAnno.s <- stat_preAnno %>%
  group_by(cnt,rep,preAnno) %>%
  summarise(count=n()) %>%
  mutate(prop= count/sum(count)) %>%
  ungroup
```


```{r  fig.height=4.8, fig.width=10.5}
pcom.preAnno <- stat_preAnno.s %>%
  ggplot(aes(x = preAnno, y = 100*prop, color=cnt)) +
  geom_bar(stat="summary", fun="mean", position = position_dodge(0.75), width = 0.58, fill="white") +
  theme_classic(base_size = 15) +
  scale_color_manual(values = c("skyblue","pink"), name="") +
  labs(title="Cell Composition, preAnno", y = "Proportion") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9.6)) +
  
  geom_point(aes(x=preAnno, y = 100*prop, color= cnt),
             position = position_dodge(0.75),
             shape=16,alpha=0.75,size=2.15,
             stroke=0.15, show.legend=F)

pcom.preAnno
```


#### sig.test          

glm.nb - anova.LRT           


```{r message=FALSE, warning=FALSE}

Sort.N <- c("CTL","CKO")

glm.nb_anova.preAnno <- list()

for(x1 in 1:2){
  for(x2 in 1:2){
    if(x2>x1){
      
      stat_preAnno.s_N <- stat_preAnno.s %>% filter(cnt %in% c(Sort.N[x1],Sort.N[x2]))
      
      total.N <- stat_preAnno.s_N %>%
        group_by(cnt, rep) %>%
        summarise(total=sum(count)) %>% ungroup() %>% as.data.frame()
      
      rownames(total.N) <- paste0(as.character(total.N$cnt),
                                  "_",
                                  as.character(total.N$rep))
      
      stat_preAnno.s_N$total <- total.N[paste0(stat_preAnno.s_N$cnt,
                                            "_",
                                            stat_preAnno.s_N$rep),"total"]
      
      glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]] <- list()
      
      for(AA in levels(stat_preAnno.s_N$preAnno)){
        glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]][[AA]] <- tryCatch({
          aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
          aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=1000)
          aaa.anova <- anova(aaa.glmnb, test = "LRT")
          aaa.anova$'Pr(>Chi)'[2]
        }, error=function(e){
        
          tryCatch({
            aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
            aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=100)
            aaa.anova <- anova(aaa.glmnb, test = "LRT")
            aaa.anova$'Pr(>Chi)'[2]
          }, error=function(e){
            
            tryCatch({
              aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
              aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=10)
              aaa.anova <- anova(aaa.glmnb, test = "LRT")
              aaa.anova$'Pr(>Chi)'[2]
            }, error = function(e){
              NA
            })     
          })
        })
      }
      glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]] <- unlist(glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]]) 
    }
  }
}


```


```{r}
glm.nb_anova.preAnno_df <- t(data.frame(Reduce(function(x,y){rbind(x,y)},glm.nb_anova.preAnno)))
rownames(glm.nb_anova.preAnno_df) <- names(glm.nb_anova.preAnno)
colnames(glm.nb_anova.preAnno_df) <- gsub("X","C",colnames(glm.nb_anova.preAnno_df))
glm.nb_anova.preAnno_df
```


```{r}
round(glm.nb_anova.preAnno_df,4)
```


## DEG               


```{r}
Idents(GEX.pure) <- "cnt"
```


```{r paged.print=FALSE}
list_sort <- list()

for(nn in levels(GEX.pure$preAnno)){
  list_sort[[nn]] <- c(nn)
}


names_sort <- names(list_sort)
data.frame(list_sort)
```


### run            

```{r}
GEX.DEGs_sort <- list()

for(nn in names_sort){
  GEX.DEGs_sort[[nn]] <- FindAllMarkers(subset(GEX.pure, subset = preAnno %in% list_sort[[nn]]),
                                       assay = "RNA",
                                       only.pos = T,
                                       min.pct = 0.05,
                                       logfc.threshold = 0.1,
                                       test.use = "MAST")
}

```


```{r}
lapply(GEX.DEGs_sort, dim)
```


```{r}
#GEX.DEGs_sort
```



```{r}
# save DEGs' table
df.DEGs_sort <- data.frame()
for(nn in names_sort){
  df.DEGs_sort <- rbind(df.DEGs_sort, data.frame(GEX.DEGs_sort[[nn]],preAnno=nn))
}

#write.csv(df.DEGs_sort, "20230927_10x_SZJ.new20270809.DEGs.CTLvsCKO.preAnno.csv")
```




### DEG stat             

```{r}
df.DEGs_sort$preAnno <- factor(df.DEGs_sort$preAnno,
                               levels = levels(GEX.pure$preAnno))
```


```{r message=FALSE, warning=FALSE, paged.print=FALSE}
cut.padj = 0.05
cut.log2FC = 0.1   
  
cut.pct1 = 0.05

df.DEGs_sort %>% filter(p_val_adj < cut.padj & 
                       abs(avg_log2FC) > cut.log2FC & 
                       pct.1 > cut.pct1) %>%
  group_by(preAnno,cluster) %>%
  summarise(up.DEGs = n()) %>% as.data.frame()
  
```

```{r message=FALSE, warning=FALSE, paged.print=FALSE, fig.width=6, fig.height=4}
df.DEGs_sort %>% filter(p_val_adj < cut.padj & 
                       abs(avg_log2FC) > cut.log2FC & 
                       pct.1 > cut.pct1) %>%
  group_by(preAnno,cluster) %>%
  summarise(up.DEGs = n()) %>% as.data.frame() %>%
  ggplot(aes(x=preAnno, y=up.DEGs, color = cluster)) +
  geom_bar(stat="summary", fun="mean", position = position_dodge(0.75), width = 0.58, fill="white") +
  geom_text(aes(label = up.DEGs),vjust=-0.21, show.legend = F, position = position_dodge(0.75) ) +
  theme_classic(base_size = 15) +
  scale_color_manual(values = c("skyblue","pink"), name="") +
  labs(title=paste0("up.DEGs stat, pct.1>0.1, padj<",cut.padj,", |log2FC|>",cut.log2FC), y = "Proportion") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9.6),
        title =element_text(size=12, face='bold'))
```



### DEG plot                


```{r}
pp.DEGs.sort <- lapply(list_sort,function(x){NA})
```


```{r}
GEX.DEGs_sort.combine <- GEX.DEGs_sort
GEX.DEGs_sort.combine <- lapply(GEX.DEGs_sort.combine, function(x){
  x[x$cluster=="CTL","avg_log2FC"] <- -x[x$cluster=="CTL","avg_log2FC"]
  x
})

```


#### Tn                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$Tn <- GEX.DEGs_sort.combine$Tn %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.1)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="Tn CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$Tn
```


#### Th.CC                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$Th.CC <- GEX.DEGs_sort.combine$Th.CC %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="Th.CC CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$Th.CC
```


#### Th1                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$Th1 <- GEX.DEGs_sort.combine$Th1 %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="Th1 CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$Th1
```


#### iNKT                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$iNKT <- GEX.DEGs_sort.combine$iNKT %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="iNKT CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$iNKT
```


#### gdT                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$gdT <- GEX.DEGs_sort.combine$gdT %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="gdT CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$gdT
```


#### ILC2                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$ILC2 <- GEX.DEGs_sort.combine$ILC2 %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="ILC2 CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$ILC2
```


#### ILC3.Ncr1                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$ILC3.Ncr1 <- GEX.DEGs_sort.combine$ILC3.Ncr1 %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="ILC3.Ncr1 CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$ILC3.Ncr1
```

#### ILC3.nn                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$ILC3.nn <- GEX.DEGs_sort.combine$ILC3.nn %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="ILC3.nn CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$ILC3.nn
```

#### ILC3.Ccr6                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$ILC3.Ccr6 <- GEX.DEGs_sort.combine$ILC3.Ccr6 %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.1)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="ILC3.Ccr6 CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$ILC3.Ccr6
```


#### Treg                                      

```{r fig.width=9, fig.height=4.5}
pp.DEGs.sort$Treg <- GEX.DEGs_sort.combine$Treg %>%
  mutate(label=ifelse(((p_val_adj < 1e-2 & avg_log2FC<0) | (p_val_adj < 1e-2 & avg_log2FC>0) | (p_val_adj < 0.05 & abs(avg_log2FC) > 0.2)) &
                        #!grepl("^Gm|^Hsp|^Rps|^Rpl|Rik$|Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
                        !grepl("Xist|Tsix|Uty|Ddx3y|Eif2s3y|Kdm5d",gene),gene,""),
         sig.=ifelse(p_val_adj<0.05 & avg_log2FC<0,"CTL",ifelse(p_val_adj<0.05 & avg_log2FC>0,"CKO","None")),
         padj=ifelse(p_val_adj<1e-60,p_val_adj+1e-60,p_val_adj)) %>%
  ggplot(aes(y= -log10(padj), x=avg_log2FC, size=pct.1, label=label,fill=sig.)) +
  geom_point(shape=21, alpha=0.65) +
  geom_text_repel(size=2.5, max.overlaps = 500) +
  scale_fill_manual(values = c("CTL"="Skyblue",
                               "CKO"="pink",
                               "None"="grey")) +
  theme_classic() + labs(title="Treg CTL vs CKO") +
  guides(size = guide_legend(order = 2), fill=guide_legend(order = 1, reverse = F)) +
  geom_vline(xintercept = c(-0.1,0.1), linetype="dotted") #+ xlim(c(-0.85,0.85))
pp.DEGs.sort$Treg
```


## ILCs only            


```{r}
GEX.sub <- subset(GEX.pure, subset = preAnno %in% c("ILC2","ILC2.Ifit1",
                                                    "ILC3.CC","ILC3.Ncr1","ILC3.nn","ILC3.Ccr6"))
GEX.sub
```


```{r}
GEX.sub$preAnno <- as.character(GEX.sub$preAnno)
GEX.sub$preAnno[GEX.sub$preAnno %in% c("ILC2.Ifit1")] <- "ILC2"

GEX.sub$preAnno <- factor(GEX.sub$preAnno,
                          levels = c("ILC2",
                                     "ILC3.CC","ILC3.Ncr1","ILC3.nn","ILC3.Ccr6"))

```



```{r fig.width=10.5,fig.height=4.5}
DimPlot(GEX.sub, reduction = "umap", label = T, group.by = "preAnno") +
  DimPlot(GEX.sub, reduction = "umap", label = F, group.by = "cnt", cols = color.cnt)
```


```{r fig.width=7.2, fig.height=5.4}
cowplot::plot_grid(
pheatmap::pheatmap(table(FB.info=GEX.sub$FB.info,
      clusters=GEX.sub$preAnno)[c(3:10),],
                   main = "Cell Count",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.0f", legend = F, silent = T)$gtable,

pheatmap::pheatmap(100*rowRatio(table(FB.info=GEX.sub$FB.info,
                                clusters=GEX.sub$preAnno)[c(3:10),]),
                   main = "Cell Ratio",
                   gaps_row = c(4),
      #gaps_col = c(12,13,14),
                   cluster_rows = F, cluster_cols = F, display_numbers = T, number_format = "%.1f", legend = F, silent =T)$gtable,
ncol = 1)
```


```{r}
stat_preAnno <- GEX.sub@meta.data[,c("cnt","rep","preAnno")]

stat_preAnno.s <- stat_preAnno %>%
  group_by(cnt,rep,preAnno) %>%
  summarise(count=n()) %>%
  mutate(prop= count/sum(count)) %>%
  ungroup
```


```{r  fig.height=3.8, fig.width=7.5}
pcom.preAnno <- stat_preAnno.s %>%
  ggplot(aes(x = preAnno, y = 100*prop, color=cnt)) +
  geom_bar(stat="summary", fun="mean", position = position_dodge(0.75), width = 0.58, fill="white") +
  theme_classic(base_size = 15) +
  scale_color_manual(values = c("skyblue","pink"), name="") +
  labs(title="Cell Composition, preAnno", y = "Proportion") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 9.6)) +
  
  geom_point(aes(x=preAnno, y = 100*prop, color= cnt),
             position = position_dodge(0.75),
             shape=16,alpha=0.75,size=2.15,
             stroke=0.15, show.legend=F)

pcom.preAnno
```


#### sig.test          

glm.nb - anova.LRT           


```{r message=FALSE, warning=FALSE}

Sort.N <- c("CTL","CKO")

glm.nb_anova.preAnno <- list()

for(x1 in 1:2){
  for(x2 in 1:2){
    if(x2>x1){
      
      stat_preAnno.s_N <- stat_preAnno.s %>% filter(cnt %in% c(Sort.N[x1],Sort.N[x2]))
      
      total.N <- stat_preAnno.s_N %>%
        group_by(cnt, rep) %>%
        summarise(total=sum(count)) %>% ungroup() %>% as.data.frame()
      
      rownames(total.N) <- paste0(as.character(total.N$cnt),
                                  "_",
                                  as.character(total.N$rep))
      
      stat_preAnno.s_N$total <- total.N[paste0(stat_preAnno.s_N$cnt,
                                            "_",
                                            stat_preAnno.s_N$rep),"total"]
      
      glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]] <- list()
      
      for(AA in levels(stat_preAnno.s_N$preAnno)){
        glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]][[AA]] <- tryCatch({
          aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
          aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=1000)
          aaa.anova <- anova(aaa.glmnb, test = "LRT")
          aaa.anova$'Pr(>Chi)'[2]
        }, error=function(e){
        
          tryCatch({
            aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
            aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=100)
            aaa.anova <- anova(aaa.glmnb, test = "LRT")
            aaa.anova$'Pr(>Chi)'[2]
          }, error=function(e){
            
            tryCatch({
              aaa <- stat_preAnno.s_N %>% filter(preAnno==AA)
              aaa.glmnb <- MASS::glm.nb(count ~ cnt + offset(log(total)), data = aaa, maxit=10)
              aaa.anova <- anova(aaa.glmnb, test = "LRT")
              aaa.anova$'Pr(>Chi)'[2]
            }, error = function(e){
              NA
            })     
          })
        })
      }
      glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]] <- unlist(glm.nb_anova.preAnno[[paste0(Sort.N[x1],"vs",Sort.N[x2])]]) 
    }
  }
}


```


```{r}
glm.nb_anova.preAnno_df <- t(data.frame(Reduce(function(x,y){rbind(x,y)},glm.nb_anova.preAnno)))
rownames(glm.nb_anova.preAnno_df) <- names(glm.nb_anova.preAnno)
colnames(glm.nb_anova.preAnno_df) <- gsub("X","C",colnames(glm.nb_anova.preAnno_df))
glm.nb_anova.preAnno_df
```


```{r}
round(glm.nb_anova.preAnno_df,5)
```














































